File: leetcode/98-validate-binary-search-tree/main.go
package main

import (
	"math"
)

/*
	Questions to ask:
	- what if there are duplicate node.vals? if there are duplicate values, it is invalid
	- what is the integraer range of node.val? -2^32 -> 2^32-1 ?

	Follow up:
	- find the invalid node in a BST
*/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	1st approach
	- compare the node.val with min & max in each recursion
	Time 	O(n)
	Space	O(h)
	4ms beats 100%
	14jan2019
*/
func isValidBST(root *TreeNode) bool {
	return dfs(root, math.MinInt64, math.MaxInt64)
}

func dfs(node *TreeNode, min int, max int) bool {
	if node == nil {
		return true
	}
	if node.Val <= min || node.Val >= max {
		return false
	}
	return dfs(node.Left, min, node.Val) && dfs(node.Right, node.Val, max)
}

/*
	2nd approach
	- compare the node.val with min & max in each iteration
	8ms beats 100%
	14jan2019
*/
type Stack struct {
	Node *TreeNode
	Min  int
	Max  int
}

func isValidBST1(root *TreeNode) bool {
	if root == nil {
		return true
	}
	stack := []Stack{} // i dont need to check if the stack is nil, so no need to use pointer
	stack = append(stack, Stack{root, math.MinInt64, math.MaxInt64})
	for len(stack) > 0 {
		pop := stack[len(stack)-1]
		curNode := pop.Node
		stack = stack[:len(stack)-1]
		if curNode.Val <= pop.Min || curNode.Val >= pop.Max {
			return false
		}
		if curNode.Left != nil {
			stack = append(stack, Stack{curNode.Left, pop.Min, curNode.Val})
		}
		if curNode.Right != nil {
			stack = append(stack, Stack{curNode.Right, curNode.Val, pop.Max})
		}
	}
	return true
}

func main() {
}

File: leetcode/98-validate-binary-search-tree/main.py
import sys

# Definition for a binary tree node.


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


"""
    1st approach
    - compare the node.val with min & max in each recursion
    
    Time 	O(N)
    Space	O(N)
    32 ms, faster than 99.46%
"""


class Solution(object):
    def isValidBST(self, root):
        return self.isValid(root, -sys.maxsize, sys.maxsize)

    def isValid(self, node, lower, upper):
        if node == None:
            return True
        if node.val <= lower or node.val >= upper:
            return False
        if self.isValid(node.left, lower, node.val) == False:
            return False
        return self.isValid(node.right, node.val, upper)


"""
    2nd approach: stack based dfs
    - compare the node.val with min & max in each iteration

    Time 	O(n)
    Space	O(h)
     44 ms, faster than 67.43%
"""


class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root == None:
            return True
        stack = []
        stack.append((root, -sys.maxsize, sys.maxsize))
        while len(stack) > 0:
            node, left, right = stack.pop()
            if node.val <= left or node.val >= right:
                return False
            if node.left != None:
                stack.append((node.left, left, node.val))
            if node.right != None:
                stack.append((node.right, node.val, right))
        return True


"""
    followup: find out the incorrect node
"""


class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.helper(root, -sys.maxsize, sys.maxsize)

    def helper(self, node, left, right):
        if node == None:
            return None
        if node.val <= left or node.val >= right:
            return node
        a = self.helper(node.left, left, node.val)
        b = self.helper(node.right, node.val, right)
        # we should return a node if there is one
        return a or b
